crates/codegen/runtime/generator/src/parser/codegen/scanner_definition.rs

```diff
@@ -1,132 +1,216 @@
 use std::collections::BTreeSet;
 
+use codegen_language_definition::model::{self, Identifier};
 use inflector::Inflector;
 use proc_macro2::TokenStream;
 use quote::{format_ident, quote};
 
 use crate::parser::codegen::versioned::VersionedQuote;
-use crate::parser::grammar::{ScannerDefinitionNode, ScannerDefinitionRef};
+use crate::parser::grammar::ScannerDefinition;
 
-pub trait ScannerDefinitionCodegen {
-    fn to_scanner_code(&self) -> TokenStream;
-    fn literals(&self) -> Vec<String>;
+impl ScannerDefinition for model::TriviaItem {
+    fn name(&self) -> &Identifier {
+        &self.name
+    }
+
+    fn to_scanner_code(&self) -> proc_macro2::TokenStream {
+        self.scanner.to_scanner_code()
+    }
+
+    fn literals(&self) -> Option<BTreeSet<String>> {
+        self.scanner.literals()
+    }
 }
 
-impl ScannerDefinitionCodegen for ScannerDefinitionRef {
-    fn to_scanner_code(&self) -> TokenStream {
-        self.node().to_scanner_code()
+impl ScannerDefinition for model::FragmentItem {
+    fn name(&self) -> &Identifier {
+        &self.name
     }
-    fn literals(&self) -> Vec<String> {
-        let mut result = BTreeSet::new();
-        if self.node().literals(&mut result) {
-            result.into_iter().collect()
-        } else {
-            vec![]
+
+    fn to_scanner_code(&self) -> proc_macro2::TokenStream {
+        VersionedScanner::new(&self.scanner, self.enabled.as_ref()).to_scanner_code()
+    }
+
+    fn literals(&self) -> Option<BTreeSet<String>> {
+        self.scanner.literals()
+    }
+
+    fn version_specifier(&self) -> Option<&model::VersionSpecifier> {
+        self.enabled.as_ref()
+    }
+}
+
+impl ScannerDefinition for model::TokenItem {
+    fn name(&self) -> &Identifier {
+        &self.name
+    }
+
+    fn to_scanner_code(&self) -> proc_macro2::TokenStream {
+        let defs: Vec<_> = self
+            .definitions
+            .iter()
+            .map(|def| VersionedScanner::new(&def.scanner, def.enabled.as_ref()))
+            .collect();
+
+        match defs.len() {
+            0 => panic!("Token {} has no definitions", self.name),
+            1 => defs.into_iter().next().unwrap().to_scanner_code(),
+            _ => choice_to_scanner_code(&defs),
         }
     }
+
+    fn literals(&self) -> Option<BTreeSet<String>> {
+        self.definitions
+            .iter()
+            .try_fold(BTreeSet::new(), |mut acc, def| {
+                let literals = def.scanner.literals()?;
+                acc.extend(literals);
+                Some(acc)
+            })
+    }
 }
 
-pub(super) trait ScannerDefinitionNodeCodegen {
+pub(crate) trait ScannerExt {
```

Comment:
I'm not keen on this meaningless `Ext` suffix. What role is being added to `Scanner`. Originally we started with the pattern `Scanner<XXXX>Extensions`

Reply:
Yes, I agree that the final name is unfortunate and I'll adapt the code here.

While Rust's `std` does use the `Ext` trait pattern, it's used more correctly there to extend the original functionality and here I need to establish a trait for the data type `model::Scanner` but also use it as an interface using my own extension of `VersionedScanner` not to repeat myself.

I will probably settle around `VersionedScanner` as the central type to be used in codegen and only convert from the `model::Scanner` rather than implement the codegen for it separately as well.

Reply:
Actually, going back I now realize that I did settle on this solution not to needlessly repeat myself here:
- we don't want to introduce `VersionedScanner` (codegen model, de facto `ScannerDefinitionNode` previously) as repeated variants from `model::Scanner` - this would duplicate the code, make it less maintainable and converting deeply from one data type to the other would be wasteful and unnecessary
- if we want to only extend the functionality, we still need to implement the codegen trait for `model::Scanner` itself to re-use the existing ADT variants, but we also want to reuse the logic for the extended, versioned, type

As such, I think just being consistent with the other `Codegen` types is actually better rather than the `Ext`. I will rename this to `ScannerCodegen` for the time being.